<a href="https://colab.research.google.com/github/timrocar/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [61]:
# YOUR WORK HERE
import pandas as pd
from scipy import stats
import numpy as np 
from scipy.stats import chisquare

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-14 21:46:45--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.6’

student.zip.6       100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-07-14 21:46:45 (146 KB/s) - ‘student.zip.6’ saved [20478/20478]



In [62]:
!unzip student.zip

Archive:  student.zip
replace student-mat.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [63]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [64]:
# YOUR WORK HERE
df.describe(exclude = 'number')

school  sex address famsize  ... nursery higher internet romantic
count     395  395     395     395  ...     395    395      395      395
unique      2    2       2       2  ...       2      2        2        2
top        GP    F       U     GT3  ...     yes    yes      yes       no
freq      349  208     307     281  ...     314    375      329      263

[4 rows x 17 columns]

In [65]:
dep1 = pd.crosstab(df['paid'], df['higher'])
dep2 = pd.crosstab(df['famsize'], df['famsup'])
ind1 = pd.crosstab(df['nursery'], df['sex'])
ind2 = pd.crosstab(df['romantic'], df['Mjob'])


# *Dependent 1*

1.) Null Hypothesis : Independent 
2.) Alt Hypothesis: Dependent 
3.) Confidence Level : 95%




In [66]:
## dependent 1!

chi2, p_value, dof, expected = stats.chi2_contingency(dep1)
print('chi2', chi2)
print('p value', p_value)
print('DOF', dof)
print('expected frequencies table \n', expected)

chi2 12.462689863341426
p value 0.0004151616557832046
DOF 1
expected frequencies table 
 [[ 10.83544304 203.16455696]
 [  9.16455696 171.83544304]]


4.) Conclusion
due to a chi^2 statistic of 12.46, and a p-value of .0004 , we reject the null hypothesis that paid and higher are independent


# *Dependent 2*

1.) Null Hypothesis : Independent 
2.) Alt Hypothesis: Dependent 
3.) Confidence Level : 95%


In [67]:
## dependent 2
chi2, p_value, dof, expected = stats.chi2_contingency(dep2)
print('chi2', chi2)
print('p value', p_value)
print('DOF', dof)
print('expected frequencies table \n', expected)

chi2 4.535754571021557
p value 0.03319381961068291
DOF 1
expected frequencies table 
 [[108.84303797 172.15696203]
 [ 44.15696203  69.84303797]]


4.) Conclusion
due to a chi^2 statistic of 4.535, and a p-value of .03 , we reject the null hypothesis that famsize and higher are famsup


# *Independent 1*

1.) Null Hypothesis : Independent 
2.) Alt Hypothesis: Dependent 
3.) Confidence Level : 95%

In [68]:
## independent 1
chi2, p_value, dof, expected = stats.chi2_contingency(ind1)
print('chi2', chi2)
print('p value', p_value)
print('DOF', dof)
print('expected frequencies table \n', expected)

chi2 0.001461464385507409
p value 0.969505016289164
DOF 1
expected frequencies table 
 [[ 42.65316456  38.34683544]
 [165.34683544 148.65316456]]


4.) Conclusion
due to a chi^2 statistic of .0014, and a p-value of .9695 , we fail to reject the null hypothesis that famsize and higher are famsup

# *Independent 2*

1.) Null Hypothesis : Independent 
2.) Alt Hypothesis: Dependent 
3.) Confidence Level : 95%

In [69]:
## independent 2
chi2, p_value, dof, expected = stats.chi2_contingency(ind2)
print('chi2', chi2)
print('p value', p_value)
print('DOF', dof)
print('expected frequencies table \n', expected)

chi2 2.357142442094112
p value 0.6703857658886252
DOF 4
expected frequencies table 
 [[39.2835443  22.63797468 93.88101266 68.57974684 38.61772152]
 [19.7164557  11.36202532 47.11898734 34.42025316 19.38227848]]


4.) Conclusion
due to a chi^2 statistic of 2.35, and a p-value of .6703 , we fail to reject the null hypothesis that famsize and higher are famsup

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [79]:
# YOUR WORK HERE
observed = pd.crosstab(df['paid'], df['higher'])

observed_w_margins = pd.crosstab(df['paid'], df['higher'], margins=True)

observed_w_margins

higher  no  yes  All
paid                
no      19  195  214
yes      1  180  181
All     20  375  395

In [89]:
observed.values

array([[ 19, 195],
       [  1, 180]])

In [80]:
#expected value calculation 
higher_vs_paid = 214*20/395
higher_vs_paid

10.835443037974683

In [81]:
row_totals = df['higher'].value_counts().sort_index().values

row_totals

array([ 20, 375])

In [82]:
col_totals = df['paid'].value_counts().sort_index().values

col_totals

array([214, 181])

In [85]:
sample_size = df.shape[0]

In [87]:
expected = np.array([[0.0,0.0],
                     [0.0,0.0]])

for row_position, col in enumerate(col_totals):
  for col_position, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[row_position][col_position] = expected_value

expected

array([[ 10.83544304, 203.16455696],
       [  9.16455696, 171.83544304]])

1) Null Hypothesis:

Wanting to take "higher" education and extra 'paid' classes are **independent**

2) Alternative Hypothesis:

Wanting to take "higher" education and extra 'paid' classes are**dependent**

3) Confidence Level: 95%

In [90]:
chi2 = ((observed - expected)**2 / expected).sum().sum()

chi2

14.141742396860645

DOF = (# rows_crosstab-1) * (# cols_crosstab-1)
DOF = 1

Using the chi^2 statistic to get a p-value: 
p_value = 0.00017

4.) Conclusion:

due to a chi^2 statistic of 14.14and a p-value of .00017, we rejext the null hypothesis that wanting to take 'higher' education and taking extra 'paid; classes are independent.

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [71]:
# YOUR WORK HERE




### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)